app/pages/project/affinity/AffinityPage.tsx

```diff
@@ -0,0 +1,162 @@
+import { EmptyCell, SkeletonCell } from '~/table/cells/EmptyCell'
+import { makeLinkCell } from '~/table/cells/LinkCell'
+import { Columns } from '~/table/columns/common'
+import { Table } from '~/table/Table'
+import { Badge } from '~/ui/lib/Badge'
+import { EmptyMessage } from '~/ui/lib/EmptyMessage'
+import { PageHeader, PageTitle } from '~/ui/lib/PageHeader'
+import { Slash } from '~/ui/lib/Slash'
+import { TableEmptyBox } from '~/ui/lib/Table'
+import { intersperse } from '~/util/array'
+import { ALL_ISH } from '~/util/consts'
+import { pb } from '~/util/path-builder'
+import type * as PP from '~/util/path-params'
+
+const antiAffinityGroupList = ({ project }: PP.Project) =>
+  getListQFn('antiAffinityGroupList', { query: { project, limit: ALL_ISH } })
+const memberList = ({ antiAffinityGroup, project }: PP.AntiAffinityGroup) =>
+  apiq('antiAffinityGroupMemberList', {
+    path: { antiAffinityGroup },
+    // We only need to get the first 2 members for preview
+    query: { project, limit: 2 },
+  })
+
+export async function clientLoader({ params }: LoaderFunctionArgs) {
+  const { project } = getProjectSelector(params)
+  await queryClient
+    .fetchQuery(antiAffinityGroupList({ project }).optionsFn())
+    .then((data) => {
+      // Preload the anti-affinity group details
+      data.items.forEach((antiAffinityGroup) => {
+        queryClient.fetchQuery(
+          memberList({ antiAffinityGroup: antiAffinityGroup.name, project })
+        )
+      })
+    })
+  return null
+}
+
+const colHelper = createColumnHelper<AntiAffinityGroup>()
+
+export const AffinityPageHeader = ({ name = 'Affinity' }: { name?: string }) => (
+  <PageHeader>
+    <PageTitle icon={<Affinity24Icon />}>{name}</PageTitle>
+    {/* TODO: Add a DocsPopover with docLinks.affinity once the doc page exists */}
+  </PageHeader>
+)
+
+type AffinityGroupPolicyBadgeProps = { policy: AffinityPolicy; className?: string }
+const AffinityGroupPolicyBadge = ({ policy, className }: AffinityGroupPolicyBadgeProps) => (
+  <Badge
+    color="neutral"
+    variant={{ allow: 'default' as const, fail: 'solid' as const }[policy]}
+    className={className}
+  >
+    {policy}
+  </Badge>
+)
+
+export default function AffinityPage() {
+  const { project } = useProjectSelector()
+  const staticCols = [
+    colHelper.accessor('name', {
+      cell: makeLinkCell((antiAffinityGroup) =>
+        pb.antiAffinityGroup({ project, antiAffinityGroup })
+      ),
+    }),
+    colHelper.accessor('description', Columns.description),
+    colHelper.accessor(() => {}, {
+      header: 'type',
+      cell: () => <Badge>anti-affinity</Badge>,
+    }),
+    colHelper.accessor('policy', {
+      cell: (info) => <AffinityGroupPolicyBadge policy={info.getValue()} />,
+    }),
+    colHelper.accessor('name', {
+      header: 'members',
+      cell: (info) => <AffinityGroupMembersCell antiAffinityGroup={info.getValue()} />,
+    }),
+    colHelper.accessor('timeCreated', Columns.timeCreated),
+  ]
```

Comment:
Same deal as the other one, though in this case because you have 5 static columns, I think it is worth it to pull out the true `staticCols` at top level and then spread that into the list here inside the `useMemo`.